internal/resource/tree.go

```diff
@@ -0,0 +1,164 @@
+package resource
+
+import (
+	"encoding/json"
+	"slices"
+
+	apiv1 "github.com/Azure/eno/api/v1"
+	"github.com/emirpasic/gods/v2/trees/redblacktree"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+type indexedResource struct {
+	Resource            *Resource
+	State               *apiv1.ResourceState
+	PendingDependencies map[Ref]struct{}
+	Dependents          map[Ref]*indexedResource
+}
+
+// treeBuilder is used to index a set of resources into a stateTree.
+type treeBuilder struct {
+	byRef        map[Ref]*indexedResource
+	byGroup      *redblacktree.Tree[int, []*indexedResource]
+	byDefiningGK map[schema.GroupKind]*indexedResource
```

Comment:
byDefiningGK == CRs? maybe a comment as to why we neeed both?

Reply:
will do!